presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java

```diff
@@ -141,13 +146,14 @@ public StripeReader(
     }
 
     public Stripe readStripe(
+            int stripeOrdinal,
             StripeInformation stripe,
             OrcAggregatedMemoryContext systemMemoryUsage,
             Optional<DwrfEncryptionInfo> decryptors,
             SharedBuffer sharedDecompressionBuffer)
             throws IOException
     {
-        StripeId stripeId = new StripeId(orcDataSource.getId(), stripe.getOffset());
+        StripeId stripeId = new StripeId(orcDataSource.getId(), stripe.getOffset(), stripeOrdinal);
 
         // read the stripe footer
         StripeFooter stripeFooter = readStripeFooter(stripeId, stripe, systemMemoryUsage);
```

Comment:
do we need filter out the index streams so we don't read them in `getDiskRanges()`?

Reply:
Likely not. There is logic in the `getDiskRanges` that sums up stream lengths if the stream offset is not set. If we remove index streams it will break.

Reply:
it will break if we don't include metadata. What i meant was we can avoid loading the index block from disk if it's already available. I assume the lazy logic you mentioned is going to read the one range if any stream in it is used. That may affect the range merge logic, etc. because we don't need to include indices from the read planning

Reply:
First `getDiskRanges` calculates the ranges, it doesn't merge them. Then `readDiskRanges` reads them, but it also checks whether the index streams are available in the cache, so we don't load index streams from the disk. The unit test verifies that we don't load cached pieces from the disk.


```
    private Map<StreamId, OrcInputStream> readDiskRanges(
            StripeId stripeId,
            Map<StreamId, DiskRange> diskRanges,
            OrcAggregatedMemoryContext systemMemoryUsage,
            Optional<DwrfEncryptionInfo> decryptors,
            SharedBuffer sharedDecompressionBuffer)
            throws IOException
    {
        // get index streams from the stripe meta cache
        Map<StreamId, OrcDataSourceInput> cachedInputs = getCachedIndexStreams(stripeId, diskRanges);

        // get the rest of the streams
        Map<StreamId, OrcDataSourceInput> readInputs = stripeMetadataSource.getInputs(orcDataSource,
                stripeId,
                Maps.filterKeys(diskRanges, key -> !cachedInputs.containsKey(key)),
                cacheable);

        // merge cached and read streams
        Map<StreamId, OrcDataSourceInput> streamsData = ImmutableMap.<StreamId, OrcDataSourceInput>builder()
                .putAll(cachedInputs)
                .putAll(readInputs)
                .build();
```

Reply:
The code looks like good amount of refactoring to reduce code duplication and new logic change and it is hard to follow.